### PR TITLE
Added todo directive

### PR DESCRIPTION
**Pull request by andrewfowlie**
I have added the todo directive as described in the TAP specification - https://testanything.org/tap-version-13-specification.html

The syntax is similar to the skip directive:

```
@test "todo test that passes" {
  todo "this todo is done"
  return 0
}

@test "todo test that fails" {
  todo "this todo is outstanding"
  return 1
}
```

This results in 

```
 ✓ todo test that passes (done: this todo is done)
 - todo test that fails (todo: this todo is outstanding)
   (in test file todo.tap, line 7)
     `todo "this todo is outstanding"' failed

2 tests, 0 failures, 1 todo, 1 done
```
and in TAP format:
```
1..2
ok 1 todo test that passes # done this todo is done
not ok 2 todo test that fails # todo this todo is outstanding
# (in test file todo.tap, line 7)
#   `todo "this todo is outstanding"' failed
```

### CODE REVIEW

**Comment by andrewfowlie** [comment]
If you like this I can also update docs + add some tests

**Comment by andrewfowlie** [comment]
is it useful?

**Comment by sublimino** [comment]
Hi @andrewfowlie, it looks useful!

bats is currently trying to get to a place where 1.0 can be shipped, I guess this would make sense after that release? Paging @btamayo @mbland @jasonkarns

**Comment by andrewfowlie** [comment]
Cool. The changes are pretty small and it's useful for me because e.g., Jenkins can read TAP and knows about todos, and it'll let me write tests, then write my codes to pass the tests...

I don't know much about the development plan so have no strong opinions about whether it should be in 1.0 or later.

File: libexec/bats-exec-test [hunk 1]
```diff
@@ -80,6 +80,11 @@ skip() {
   exit 0
 }
 
+BATS_TEST_TODOED=''
+todo() {
+  BATS_TEST_TODOED=${1:-1}
```
Comment (mbland) [inline libexec/bats-exec-test]: The `${1:-1}` needs to be double-quoted.

File: libexec/bats-exec-test [hunk 2]
```diff
@@ -270,14 +277,23 @@ bats_exit_trap() {
     fi
   fi
 
+  if [ -n "$BATS_TEST_TODOED" ]; then
```
Comment (mbland) [inline libexec/bats-exec-test]: Use `[[` and `]]` for all new code, please.

File: libexec/bats-exec-test [hunk 3]
```diff
@@ -270,14 +277,23 @@ bats_exit_trap() {
...
+    todoed=" # todo"
+    todone=" # done"
+    if [ "1" != "$BATS_TEST_TODOED" ]; then
```
Comment (mbland) [inline libexec/bats-exec-test]: `[[` and `]]`.

File: libexec/bats-format-tap-stream [hunk 4]
```diff
@@ -51,6 +53,22 @@ skip() {
   advance
 }
 
+todo() {
+  local reason="$1"
+  [ -z "$reason" ] || reason=": $reason"
```
Comment (mbland) [inline libexec/bats-format-tap-stream]: Use `[[` and `]]`, and expand this into an `if` statement to make it easier to scan.

**Comment by andrewfowlie** [comment]
Thanks for the comments - will make the changes soonish

**Comment by andrewfowlie** [comment]
Is there anything more I can do here?

**Comment by mbland** [comment]
Sorry for the delay. I think we were all running a little lean during August, due to vacations and such (I traveled at the beginning of the month, and organized a large-ish event at work at the end). Will do my best to loop back and take a solid look again soon.

**Comment by tterranigma** [comment]
@andrewfowlie Since you are editing the tap formatter, could you output todo tests with color? Perhaps some blue or cyan? See also #162 .

**Comment by andrewfowlie** [comment]
@tterranigma I’m afraid I’m super busy right now, unlikely I’ll do that any time soon. Feel free to jump in and add colours

**Comment by Aazhar** [comment]
Hello,
I would like to know if this feature will merged anytime soon ?
Thanks

**Comment by sublimino** [comment]
@andrewfowlie (due to the passing of time, sorry!) this branch has some conflicts, could you rebase master onto it and ensure the tests still pass?

In general LGTM, although I'd like a second opinion and to test this in anger for a while. If you'd like to test the latest (e.g. rebased) version of this @Aazhar please do so and let us know here : ]

**Comment by Aazhar** [comment]
ok so after reviewing this PR, I'm questioning the meaning of 'done' todos in the initial post.
Actually imho, all todos should be counted as todos, regardless of their supposed result.

**Comment by andrewfowlie** [comment]
I will try to rebase next week.

Done means that the todo is complete because the test passes. I think it’s useful to indicate that (although could change the wording)

**Comment by andrewfowlie** [comment]
arggh, something has gone wrong with my PR/rebase against master. There was some small complication because some of my modified files had been removed from the upstream repo, so I did some things manually. But I thought it worked (https://github.com/bats-core/bats-core/compare/master...andrewfowlie:master), but this PR seems to include all kinds of commits and changes.

Maybe it's this problem - https://stackoverflow.com/questions/18323665/pull-request-on-github-showing-commits-rebased-from-master

**Comment by jasonkarns** [comment]
I'm assuming part of the problem here is that the changes were made on the master branch instead of on a feature branch.

**Comment by jasonkarns** [comment]
As to "done" todos...

@Aazhar the TAP spec itself indicates that tests marked as TODO that _pass_ should be counted as "bonus" tests.

Personally I disagree with the TAP spec here. I believe rspec's [behavior](https://relishapp.com/rspec/rspec-core/v/3-8/docs/pending-and-skipped-examples/pending-examples#%60pending%60-any-arbitrary-reason-with-a-passing-example) is better here. A test that's marked as TODO should _not_ pass. If it does, the test is either not a valid test (it's testing the wrong thing) and needs to be updated; or the test is valid and the functionality is complete, in which case, the test needs to be updated (remove the TODO marker). Ergo, in both scenarios, the test needs "fixed". Rspec reports `pending` tests that pass as failures, and I believe that is the proper behavior. However, since the TAP spec is explicit, and at least TAP reports them separate from regular passing tests, I'll concede that spec-compliant behavior is acceptable.

However, as a test suite that contains DONE TODOs requires changes, I think we should follow rspec's lead here and return a non-zero exit for the suite in such a case.

**Comment by andrewfowlie** [comment]
That makes sense. I am very busy travelling right now. I’m happy for anyone to finish this PR/if anyone knows how to fix the aforementioned GitHub problem with the PR please let me know

**Pull request closed without merging**